From: Chad Monroe <chad@monroe.io>
Date: Mon, 19 May 2025 10:11:58 -0700
Subject: [PATCH] net: dsa: mxl862xx: add USXGMII LAN support 

add support for using the 2nd USXGMII interface as a 10G LAN port

Signed-off-by: Chad Monroe <chad@monroe.io>
---
 drivers/net/dsa/mxl862xx/mxl862xx.c |  101 ++++++++++++++++++--------
 drivers/net/dsa/mxl862xx/mxl862xx.h |    4 -
 2 files changed, 73 insertions(+), 32 deletions(-)

--- a/drivers/net/dsa/mxl862xx/mxl862xx.c
+++ b/drivers/net/dsa/mxl862xx/mxl862xx.c
@@ -43,7 +43,7 @@
 #define MXL862XX_FDMA_PCTRL_EN BIT(0) /* FDMA Port Enable */
 
 #define MAX_BRIDGES 16
-#define MAX_VLAN_ENTRIES (1024 - 160)
+#define MAX_VLAN_ENTRIES (1024 - 320)
 #define IDX_INVAL (-1)
 
 #define INGRESS_FINAL_RULES 5
@@ -190,7 +190,7 @@ static int mxl862xx_update_bridge_conf_p
 					    struct net_device *bridge, int action)
 {
 	struct mxl862xx_priv *priv = ds->priv;
-	u8 phy_ports = priv->hw_info->phy_ports;
+	u8 max_ports = priv->hw_info->max_ports;
 	u8 cpu_port = priv->hw_info->cpu_port;
 	u8 i;
 	bool vlan_sp_tag = (priv->port_info[cpu_port].tag_protocol == DSA_TAG_PROTO_MXL862_8021Q);
@@ -212,7 +212,7 @@ static int mxl862xx_update_bridge_conf_p
 
 	/* Update switch according to local bridge port map */
 	/* Add this port to the port maps of other ports skiping it's own map */
-	for (i = 0; i < phy_ports; i++) {
+	for (i = 0; i < max_ports; i++) {
 		struct mxl862xx_bridge_port_config br_port_cfg = { };
 		int bridge_id = priv->port_info[i].bridge_id;
 
@@ -2551,14 +2551,14 @@ static void mxl862xx_deisolate_port(stru
 
 	ret = mxl862xx_update_bridge_conf_port(ds, port, NULL, 0);
 	if (ret) {
-		dev_err(ds->dev, "failed to remove port %d from bridge\n", port);
+		dev_err(ds->dev, "%s: failed to remove port %d from bridge\n", __func__, port);
 		return;
 	}
 
 	bridge_alloc.bridge_id = priv->port_info[port].bridge_id;
 	ret = MXL862XX_API_WRITE(priv, MXL862XX_BRIDGE_FREE, bridge_alloc);
 	if (ret) {
-		dev_err(ds->dev, "failed to free bridge %d\n", bridge_alloc.bridge_id);
+		dev_err(ds->dev, "%s: failed to free bridge %d\n", __func__, bridge_alloc.bridge_id);
 		return;
 	}
 
@@ -2609,10 +2609,16 @@ static int mxl862xx_find_bridge_id(struc
 	struct mxl862xx_priv *priv = ds->priv;
 	u8 i;
 
-	if (bridge)
-		for (i = 0; i < priv->hw_info->phy_ports; i++)
-			if (priv->port_info[i].bridge == bridge)
-				return priv->port_info[i].bridge_id;
+	if (!bridge)
+		return 0;
+
+	for (i = 0; i < priv->hw_info->max_ports; i++) {
+		if (dsa_is_unused_port(ds, i))
+			continue;
+
+		if (priv->port_info[i].bridge == bridge)
+			return priv->port_info[i].bridge_id;
+	}
 
 	return 0;
 }
@@ -2649,17 +2655,17 @@ static void mxl862xx_set_vlan_filter_lim
 	/* The calculation of the max number of simultaneously supported VLANS (priv->max_vlans)
 	 * comes from the equation:
 	 *
-	 * MAX_VLAN_ENTRIES = phy_ports * (EGRESS_FINAL_RULES + EGRESS_VID_RULES * priv->max_vlans)
-	 *  + phy_ports * (INGRESS_FINAL_RULES + INGRESS_VID_RULES * priv-> max_vlans)
+	 * MAX_VLAN_ENTRIES = user_ports * (EGRESS_FINAL_RULES + EGRESS_VID_RULES * priv->max_vlans)
+	 *  + user_ports * (INGRESS_FINAL_RULES + INGRESS_VID_RULES * priv-> max_vlans)
 	 *  + cpu_ingress_entries + cpu_egress_entries
 	 */
 	if (priv->port_info[cpu_port].tag_protocol == DSA_TAG_PROTO_MXL862_8021Q) {
-		priv->max_vlans = (MAX_VLAN_ENTRIES - priv->hw_info->phy_ports *
+		priv->max_vlans = (MAX_VLAN_ENTRIES - priv->hw_info->user_ports *
 				   (EGRESS_FINAL_RULES + INGRESS_FINAL_RULES + 2) - 3) /
-				  (priv->hw_info->phy_ports *
+				  (priv->hw_info->user_ports *
 				   (EGRESS_VID_RULES + INGRESS_VID_RULES) + 2);
 		/* 2 entries per port and 1 entry for fixed rule */
-		cpu_ingress_entries = priv->hw_info->phy_ports * 2 + 1;
+		cpu_ingress_entries = priv->hw_info->user_ports * 2 + 1;
 		/* 2 entries per each vlan and 2 entries for fixed rules */
 		cpu_egress_entries = priv->max_vlans * 2 + 2;
 
@@ -2671,9 +2677,9 @@ static void mxl862xx_set_vlan_filter_lim
 		user_ingress_entries = INGRESS_FINAL_RULES + INGRESS_VID_RULES * priv->max_vlans;
 		user_egress_entries = EGRESS_FINAL_RULES + EGRESS_VID_RULES * priv->max_vlans;
 	} else {
-		priv->max_vlans = (MAX_VLAN_ENTRIES - priv->hw_info->phy_ports *
+		priv->max_vlans = (MAX_VLAN_ENTRIES - priv->hw_info->user_ports *
 				(EGRESS_FINAL_RULES + INGRESS_FINAL_RULES) - 1) /
-			(priv->hw_info->phy_ports * (EGRESS_VID_RULES + INGRESS_VID_RULES) + 2);
+			(priv->hw_info->user_ports * (EGRESS_VID_RULES + INGRESS_VID_RULES) + 2);
 		/* 1 entry for fixed rule */
 		cpu_ingress_entries =  1;
 		/* 2 entries per each vlan  */
@@ -2694,7 +2700,11 @@ static void mxl862xx_set_vlan_filter_lim
 		vlan->egress_vlan_block_info.filters_max - 1;
 
 	/* Set limits and indexes required for processing VLAN rules for user ports */
-	for (i = 0; i < priv->hw_info->phy_ports; i++) {
+	for (i = 0; i < priv->hw_info->max_ports; i++) {
+		//if (!dsa_is_user_port(ds, i))
+		if (dsa_is_unused_port(ds, i))
+			continue;
+
 		vlan = &priv->port_info[i].vlan;
 		vlan->ingress_vlan_block_info.entries_per_vlan = INGRESS_VID_RULES;
 		vlan->ingress_vlan_block_info.filters_max = user_ingress_entries;
@@ -2706,9 +2716,9 @@ static void mxl862xx_set_vlan_filter_lim
 		vlan->egress_vlan_block_info.final_filters_idx =
 			vlan->egress_vlan_block_info.filters_max - 1;
 	}
-	dev_info(ds->dev, "%s: phy_ports:%d, priv->max_vlans: %d, cpu_egress_entries: %d, "
+	dev_info(ds->dev, "%s: user_ports: %d, priv->max_vlans: %d, cpu_egress_entries: %d, "
 		 "user_ingress_entries: %d, INGRESS_VID_RULES: %d\n",
-		 __func__, priv->hw_info->phy_ports, priv->max_vlans,
+		 __func__, priv->hw_info->user_ports, priv->max_vlans,
 		 cpu_egress_entries, user_ingress_entries, INGRESS_VID_RULES);
 }
 
@@ -2843,14 +2853,14 @@ static void mxl862xx_port_bridge_leave(s
 	bridge_id = mxl862xx_find_bridge_id(ds, bridge.dev);
 	ret = mxl862xx_update_bridge_conf_port(ds, port, bridge.dev, 0);
 	if (ret) {
-		dev_err(ds->dev, "failed to remove port %d from bridge\n", port);
+		dev_err(ds->dev, "%s: failed to remove port %d from bridge\n", __func__, port);
 		return;
 	}
 
 	if (priv->bridge_portmap[bridge_id] == BIT(DSA_MXL_PORT(cpu_port))) {
 		ret = MXL862XX_API_WRITE(priv, MXL862XX_BRIDGE_FREE, bridge_alloc);
 		if (ret) {
-			dev_err(ds->dev, "failed to free bridge %d\n", bridge_alloc.bridge_id);
+			dev_err(ds->dev, "%s: failed to free bridge %d\n", __func__, bridge_alloc.bridge_id);
 			return;
 		}
 	}
@@ -2885,6 +2895,7 @@ static int mxl862xx_phy_write_c45_mii_bu
 
 static int mxl862xx_setup_mdio(struct dsa_switch *ds)
 {
+	struct mxl862xx_priv *priv = ds->priv;
 	struct device *dev = ds->dev;
 	struct mii_bus *bus;
 	static int idx;
@@ -2905,6 +2916,10 @@ static int mxl862xx_setup_mdio(struct ds
 	bus->parent = dev;
 	bus->phy_mask = ~ds->phys_mii_mask;
 
+	/* 10G ports do not support slave MDIO bus yet */
+	if (priv->hw_info->ext_ports <= 2)
+		bus->phy_mask |= 0xff00;
+
 #if IS_ENABLED(CONFIG_OF_MDIO)
 	struct device_node *mdio_np;
 
@@ -2968,10 +2983,16 @@ static int mxl862xx_setup(struct dsa_swi
 	for (i = 0; i < MAX_VLANS; i++)
 		priv->port_info[cpu_port].vlan.egress_vlan_block_info.vlans[i].untagged = true;
 
-	for (i = 0; i < priv->hw_info->phy_ports; i++) {
+	for (i = 0; i < priv->hw_info->max_ports; i++) {
+		if (dsa_is_unused_port(ds, i))
+			continue;
+
 		priv->port_info[i].vlan.filtering_mode_locked = false;
 		priv->port_info[i].isolated = true;
 
+		if (dsa_is_cpu_port(ds, i))
+			continue;
+
 		mxl862xx_port_state(ds, i, false);
 		mxl862xx_isolate_port(ds, i);
 		mxl862xx_port_fast_age(ds, i);
@@ -3028,10 +3049,14 @@ static void mxl862xx_phylink_get_caps(st
 
 	if (port >= 0 && port < priv->hw_info->phy_ports) {
 		__set_bit(PHY_INTERFACE_MODE_INTERNAL, config->supported_interfaces);
-	} else if (port == 8 || port == 9) {
+	} else if (port >= priv->hw_info->cpu_port && priv->hw_info->ext_ports == 2) {
+		__set_bit(PHY_INTERFACE_MODE_SGMII, config->supported_interfaces);
+		__set_bit(PHY_INTERFACE_MODE_1000BASEX, config->supported_interfaces);
+		__set_bit(PHY_INTERFACE_MODE_2500BASEX, config->supported_interfaces);
+		__set_bit(PHY_INTERFACE_MODE_10GBASER, config->supported_interfaces);
 		__set_bit(PHY_INTERFACE_MODE_USXGMII, config->supported_interfaces);
 		config->mac_capabilities |= MAC_5000FD | MAC_10000FD;
-	} else if (port > 9) {
+	} else {
 		__set_bit(PHY_INTERFACE_MODE_NA, config->supported_interfaces);
 	}
 }
@@ -3044,6 +3069,7 @@ static void mxl862xx_phylink_mac_config(
 	switch (state->interface) {
 	case PHY_INTERFACE_MODE_INTERNAL:
 	case PHY_INTERFACE_MODE_SGMII:
+	case PHY_INTERFACE_MODE_10GBASER:
 	case PHY_INTERFACE_MODE_USXGMII:
 		return;
 	default:
@@ -3256,7 +3282,6 @@ static void mxl862xx_port_mirror_del(str
 		.mask = MXL862XX_CTP_PORT_CONFIG_LOOPBACK_AND_MIRROR,
 	};
 	struct mxl862xx_priv *priv = ds->priv;
-	u8 phy_ports = priv->hw_info->phy_ports;
 	int active_mirrors = 0;
 	int ret, i;
 
@@ -3271,10 +3296,14 @@ static void mxl862xx_port_mirror_del(str
 		return;
 	}
 
-	for (i = 0; i < phy_ports; i++)
+	for (i = 0; i < priv->hw_info->max_ports; i++) {
+		if (dsa_is_unused_port(ds, i))
+			continue;
+
 		if (priv->port_info[i].egress_mirror_enabled ||
 		    priv->port_info[i].egress_mirror_enabled)
 			active_mirrors = 1;
+	}
 
 	if (!active_mirrors) {
 		struct mxl862xx_monitor_port_cfg monitor_port = { };
@@ -3298,7 +3327,10 @@ static int mxl862xx_port_fdb_add(struct
 
 	memcpy(param.mac, addr, ETH_ALEN);
 
-	for (i = 0; i < priv->hw_info->phy_ports; i++) {
+	for (i = 0; i < priv->hw_info->max_ports; i++) {
+		if (dsa_is_unused_port(ds, i))
+			continue;
+
 		if (!dsa_is_cpu_port(ds, port) && i != port)
 			continue;
 
@@ -3328,7 +3360,10 @@ static int mxl862xx_port_fdb_del(struct
 
 	memcpy(param.mac, addr, ETH_ALEN);
 
-	for (i = 0; i < priv->hw_info->phy_ports; i++) {
+	for (i = 0; i < priv->hw_info->max_ports; i++) {
+		if (dsa_is_unused_port(ds, i))
+			continue;
+
 		if (!dsa_is_cpu_port(ds, port) && i != port)
 			continue;
 
@@ -3547,14 +3582,18 @@ static void mxl862xx_remove(struct mdio_
 }
 
 static const struct mxl862xx_hw_info mxl86282_data = {
-	.max_ports = 9,
+	.max_ports = MAX_PORTS,
+	.user_ports = 9,
 	.phy_ports = 8,
+	.ext_ports = 2,
 	.cpu_port = 8,
 };
 
 static const struct mxl862xx_hw_info mxl86252_data = {
-	.max_ports = 9,
+	.max_ports = MAX_PORTS,
+	.user_ports = 6,
 	.phy_ports = 5,
+	.ext_ports = 2,
 	.cpu_port = 8,
 };
 
--- a/drivers/net/dsa/mxl862xx/mxl862xx.h
+++ b/drivers/net/dsa/mxl862xx/mxl862xx.h
@@ -1,11 +1,13 @@
 #define VID_RULES 2
 #define MAX_VLANS  100
-#define MAX_PORTS 13
+#define MAX_PORTS 15
 #define MAX_BRIDGES 16
 
 struct mxl862xx_hw_info {
 	u8 max_ports;
+	u8 user_ports;
 	u8 phy_ports;
+	u8 ext_ports;
 	u8 cpu_port;
 };
 
